hs-bindgen-runtime/src/HsBindgen/Runtime/CEnum/General.hs

```diff
@@ -0,0 +1,305 @@
+generalCEnumEnumFromThen :: forall a. GeneralCEnum a => a -> a -> [a]
+generalCEnumEnumFromThen x y
+    | i == j    = throw $ CEnumFromEqThen (fromIntegral i)
+    | otherwise = case List.dropWhile (/= i) values of
+        _i:rest -> case first ((+ 1) . length) (List.break (== j) rest) of
+          (_w, []) -> throw $ CEnumInvalid (fromIntegral j)
+          (w,  js) -> x : map (toGeneralCEnum . NonEmpty.head) (chunksOf w js)
+        []      -> throw $ CEnumInvalid (fromIntegral i)
+  where
+    i, j :: GeneralCEnumZ a
+    i = fromGeneralCEnum x
+    j = fromGeneralCEnum y
+
+    values :: [GeneralCEnumZ a]
+    values = (if i < j then id else reverse) $
+      generalCEnumValues (Proxy :: Proxy a)
+
+-- | Create a list of enumeration values like @[x .. z]@
+--
+-- This is an implementation of 'enumFromTo'.
+--
+-- This is a partial function.  A 'CEnumException' is thrown if a passed value
+-- is invalid.
+generalCEnumEnumFromTo :: forall a. GeneralCEnum a => a -> a -> [a]
+generalCEnumEnumFromTo x z =
+    case List.dropWhile (/= i) values of
+      _i:rest -> case List.break (== k) rest of
+        (js,  _k:_) -> x : map toGeneralCEnum js ++ [z]
+        (_js, [])   -> throw $ CEnumInvalid (fromIntegral k)
+      []      -> throw $ CEnumInvalid (fromIntegral i)
+  where
+    i, k :: GeneralCEnumZ a
+    i = fromGeneralCEnum x
+    k = fromGeneralCEnum z
+
+    values :: [GeneralCEnumZ a]
+    values = generalCEnumValues (Proxy :: Proxy a)
+
+-- | Create a list of enumeration values like @[x, y .. z]@
+--
+-- This is an implementation of 'enumFromThenTo'.
+--
+-- This is a partial function.  A 'CEnumException' is thrown if a passed value
+-- is invalid or if the from and then values are equal.
+generalCEnumEnumFromThenTo :: forall a. GeneralCEnum a => a -> a -> a -> [a]
+generalCEnumEnumFromThenTo x y z
+    | i == j    = throw $ CEnumFromEqThen (fromIntegral i)
+    | otherwise = case List.dropWhile (/= i) values of
+        _i:rest -> case List.break (== k) rest of
+          (is,  _k:_) ->
+            case first ((+ 1) . length) (List.break (== j) (is++[k])) of
+              (_w, []) -> throw $ CEnumInvalid (fromIntegral j)
+              (w,  js) ->
+                x : map (toGeneralCEnum . NonEmpty.head) (chunksOf w js)
+          (_js, [])   -> throw $ CEnumInvalid (fromIntegral k)
+        []      -> throw $ CEnumInvalid (fromIntegral i)
+  where
+    i, j, k :: GeneralCEnumZ a
+    i = fromGeneralCEnum x
+    j = fromGeneralCEnum y
+    k = fromGeneralCEnum z
+
+    values :: [GeneralCEnumZ a]
+    values = (if i < j then id else reverse) $
+      generalCEnumValues (Proxy :: Proxy a)
+
+{-------------------------------------------------------------------------------
+  Instance Deriving
+-------------------------------------------------------------------------------}
+
+-- | Type used to derive 'Bounded' and 'Enum' instances when a type has a
+-- 'GeneralCEnum' instance
+newtype GenCEnum a = GenCEnum a
```

Comment:
Let's rename this to `AsCEnum`, and explicitly mention "deriving-via" in the comments.